### PR TITLE
Fix race condition during debugging restart

### PR DESCRIPTION
**Pull request by ericdevg**
#142

@AndreSayreMSFT @HuanhuanSunMSFT

### CODE REVIEW

**Comment by ASayre** [comment]
Build failed?

File: PowerShellTools/DebugEngine/ScriptDebugger.cs [hunk 1]
```diff
@@ -262,19 +284,27 @@ private void RefreshCallStack()
...
             try
             {
-                result = DebuggingService.GetCallStack();
+                if (IsDebugging)
+                {
+                    result = DebuggingService.GetCallStack();
+                }
+                else
+                {
+                    throw new Exception("Not in debugging mode any more!");
```
Comment (ASayre) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: We should throw a more specific exception if one exists; but if not, these can be bubbled up to the user and should be loc'd.

File: PowerShellTools/DebugEngine/ScriptDebugger.cs [hunk 2]
```diff
@@ -219,7 +236,11 @@ public void UpdateBreakpoint(DebuggerBreakpointUpdatedEventArgs e)
         /// </summary>
         public void DebuggerFinished()
         {
-            IsDebuggingCommandReady = false;
+            lock (_debuggingCmdLock)
+            {
+                IsDebuggingCommandReady = false;
```
Comment (ASayre) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: Root cause of the race condition?  I think I'm missing something as far as what was being hit X-thread, and is fixed up with the lock around IsDebuggingCommandReady

**Comment by ASayre** [comment]
Tried build again, and it's halting on ShouldBreakOnBreakPoint.

File: PowerShellTools/DebugEngine/ScriptDebugger.cs (hunk 2 above)
Comment (ericdevg) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: Talked offline, this part is just for thread-safe purpose, not specifically for this race condition.

File: PowerShellTools/DebugEngine/ScriptDebugger.cs (hunk 1 above)
Comment (ericdevg) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: I think I was just a bit lazy that trying to use this exception as a fast-fail for this method, and meanwhile leveraging it as a return value to its consumer, so this ex is completely internal and string won't be user-visible. But I agree with you that needs to be more specific exception.

Comment (ericdevg) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: And I m making it more specific.

File: PowerShellTools/DebugEngine/DebugEngineInternalException.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PowerShellTools.DebugEngine
+{
+    class DebugEngineInternalException: Exception
```
Comment (ASayre) [inline PowerShellTools/DebugEngine/DebugEngineInternalException.cs]: access modifier

File: PowerShellTools/DebugEngine/ScriptDebugger.cs (hunk 2 above)
Comment (ASayre) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: But does the lock do anything here?  we don't read/use any values within these lock blocks, as far as I can tell, the only thing contained within them is setting IsDebuggingCommandReady .  Eg, one thread can set it to true; and another could set it to false right after, when we're off of the critical path...

Comment (ericdevg) [inline PowerShellTools/DebugEngine/ScriptDebugger.cs]: Yes, those all becomes atomic ops, we don't need those locks. Removing them.

**Comment by ASayre** [comment]
Looks good!

**Pull request merged**